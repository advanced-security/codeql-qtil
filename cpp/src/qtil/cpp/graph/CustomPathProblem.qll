private import qtil.locations.CustomPathProblem
private import qtil.cpp.locations.Locatable
private import cpp
// Import the C++ specific configuration for making custom path problems.
import PathProblem<Location, CppLocatableConfig>
