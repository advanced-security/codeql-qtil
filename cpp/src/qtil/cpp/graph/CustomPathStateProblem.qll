private import qtil.locations.CustomPathStateProblem
private import qtil.cpp.locations.Locatable
private import cpp
// Import the C++ specific configuration for making custom path state problems.
import PathStateProblem<Location, CppLocatableConfig>
