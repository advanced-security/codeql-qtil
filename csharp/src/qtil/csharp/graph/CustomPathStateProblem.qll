private import qtil.locations.CustomPathStateProblem
private import qtil.csharp.locations.Locatable
private import csharp
// Import the C# specific configuration for making custom path state problems.
import PathStateProblem<Location, CsharpLocatableConfig>
