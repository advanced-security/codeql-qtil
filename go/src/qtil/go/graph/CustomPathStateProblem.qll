private import qtil.locations.CustomPathStateProblem
private import qtil.go.locations.Locatable
private import go
// Import the Go specific configuration for making custom path state problems.
import PathStateProblem<go::DbLocation, GoLocatableConfig>
