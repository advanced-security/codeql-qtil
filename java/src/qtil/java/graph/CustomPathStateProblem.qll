private import qtil.locations.CustomPathStateProblem
private import qtil.java.locations.Locatable
private import java
// Import the Java specific configuration for making custom path state problems.
import PathStateProblem<Location, JavaLocatableConfig>
