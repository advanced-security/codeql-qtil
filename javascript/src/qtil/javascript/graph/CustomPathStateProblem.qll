private import qtil.locations.CustomPathStateProblem
private import qtil.javascript.locations.Locatable
private import javascript
// Import the javascript specific configuration for making custom path state problems.
import PathStateProblem<DbLocation, JavascriptLocatableConfig>