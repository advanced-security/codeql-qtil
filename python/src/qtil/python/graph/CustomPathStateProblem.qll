private import qtil.locations.CustomPathStateProblem
private import qtil.python.locations.Locatable
private import python
// Import the python specific configuration for making custom path state problems.
import PathStateProblem<Location, PythonLocatableConfig>
