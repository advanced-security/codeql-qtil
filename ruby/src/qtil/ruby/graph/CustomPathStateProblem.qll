private import qtil.locations.CustomPathStateProblem
private import qtil.ruby.locations.Locatable
private import ruby
// Import the Ruby specific configuration for making custom path state problems.
import PathStateProblem<Location, RubyLocatableConfig>