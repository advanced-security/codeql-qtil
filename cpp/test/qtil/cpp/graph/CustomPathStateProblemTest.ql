// Minimal test to make sure CustomPathStateProblem works with C++ code.
import cpp
import qtil.cpp.graph.CustomPathStateProblem

module CallGraphPathStateProblemConfig implements CustomPathStateProblemConfigSig {
  class Node = Variable;

  class State = int; // Track search depth

  predicate start(Node n, int depth) { n.getName() = "start" and depth = 0 }

  bindingset[depth]
  predicate end(Node n, int depth) { n.getName() = "end" and depth >= 0 }

  bindingset[depth1]
  bindingset[depth2]
  predicate edge(Variable a, int depth1, Variable b, int depth2) {
    depth2 = depth1 + 1 and
    exists(Initializer initA |
      a.getInitializer() = initA and
      initA.getExpr().(VariableAccess).getTarget() = b
    )
  }
}

import CustomPathStateProblem<CallGraphPathStateProblemConfig>

from Variable start, Variable end
where problem(start, end)
select start, end, "Path from $@ to $@.", start.getName(), start, end.getName(), end
