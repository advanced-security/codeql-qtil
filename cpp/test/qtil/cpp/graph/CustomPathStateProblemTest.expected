edges
| test.cpp:3:9:3:11 | mid | test.cpp:2:9:2:11 | end |  |  |
| test.cpp:4:9:4:13 | start | test.cpp:3:9:3:11 | mid |  |  |
nodes
| test.cpp:2:9:2:11 | end | semmle.label | end |
| test.cpp:3:9:3:11 | mid | semmle.label | mid |
| test.cpp:4:9:4:13 | start | semmle.label | start |
#select
| test.cpp:4:9:4:13 | start | test.cpp:2:9:2:11 | end | Path from $@ to $@. | start | test.cpp:4:9:4:13 | start | end | test.cpp:2:9:2:11 | end |